README
% yaft
% haru
% Last update: 12/11/13

yet another framebuffer terminal

## download
-	[yaft-0.2.0](http://www.nak.ics.keio.ac.jp/~haru/yaft/release/yaft-0.2.0.tar.gz)
-	[yaft-freebsd-0.1.9](http://www.nak.ics.keio.ac.jp/~haru/yaft/release/yaft-freebsd-0.1.9.tar.gz)

## features
+	recognizes most of escape sequences of vt102 and linux console ([all sequences](escape.html))
+	supports various framebuffer types (8/15/16/24/32bpp)
+	supports (only) UTF-8 encoding
+	supports 256 colors (same as xterm)
+	supports wallpaper

## recent changes
for more detail, see [ChangeLog](./release/yaft-current/ChangeLog)

### version 0.1.9 (2012-10-18)

-	supported virtual console switching
-	removed all optios without "YAFT=wall"

### version 0.2.0 (2012-11-09)

-	suppported BDF

~~~
	$ ./mkfont alias /your/favorite/bdf > glyph.h
	$ make yaft
~~~

## screenshot
![](http://www.nak.ics.keio.ac.jp/~haru/yaft/img/yaft-screenshot.png)

<!--
## configuration
if you want to change configuration, rewrite conf.h

### path and terminal name

+	static char *fb_path = "/dev/fb0";
+	static char *shell_cmd = "/bin/bash";
+	static char *term_name = "TERM=yaft-256color";

### color
the value is an index of color_list[] in color.h

+	DEFAULT_FG = 7,
+	DEFAULT_BG = 0,
+	CURSOR_COLOR = 2,

### misc

+	DEBUG = false,
+	TABSTOP = 8,
+	SUBSTITUTE_HALF = 0x20, /* SPACE */
+	SUBSTITUTE_WIDE = 0x3000, /* IDEOGRAPHIC SPACE */

## install

~~~
$ make
# make install
~~~

## usage

~~~
$ yaft
~~~

for displaying wallpaper, you can use yaft_wall script (it requires [fbv])

~~~
$ yaft_wall /path/to/wallpaper.jpg
~~~

[fbv]: http://www.eclis.ch/fbv/

## tools
some useful tools (bdf to yaft font converter etc) are found in tools/ directory

## license
Copyright (c) 2012 haru (uobikiemukot at gmail dot com)

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
-->


ChangeLog
## 2012-06-19
-	version 0.1.0
-	initial release

## 2012-06-19
-	version 0.1.1
-	bug fix

## 2012-06-20
-	version 0.1.2
-	bug fix

## 2012-06-20
-	version 0.1.3
-	bug fix

## 2012-06-21
-	version 0.1.4
-	add eat_newline_glitch (xel)
-	remove auto_right_margin (bw)
-	remove OSC 4 (ccc, initc)

## 2012-06-24
-	version 0.1.5
-	change fonts format
-	fix glyph drawing
-	change effects of BOLD and REVERSE

## 2012-06-29
-	version 0.1.6
-	support more than one font
-	support glyph alias

## 2012-10-06
-	version 0.1.7
-	support following framebuffer types:
	-	15/16/24/32bpp true color (and maybe direct color)
	-	8bpp pseudo color

## 2012-10-17
-	version 0.1.8
-	add some opitons
	-	yaft reads environment value, YAFT
		-	wall: display wallpaper
		-	w=N: set terminal width (pixel)
		-	h=N: set terminal height (pixel)
		-	x=N: set x offset
		-	y=N: set y offset

~~~
$ export YAFT="wall w=640 h=384 x=32 y=32"; yaft
~~~

## 2012-10-18
-	version 0.1.9
-	add some codes related to virtual console
-	remove all optios without "YAFT=wall"

## 2012-11-09
-	version 0.2.0
-	support BDF

~~~
$ ./mkfont alias-file your/favorite/bdf > glyph.h
$ make yaft
~~~

## 2012-11-30
-	version 0.2.1
-	bug fix
	-	single CSI u (not following CSI s) causes Segmentation fault (reported by saitoha ([@kefir_]))

[@kefir_]: http://saitoha.github.com/

## 2012-12-02
-	version 0.2.2
-	improve UTF-8 parser (valid for [UTF-8 decoder capability and stress test])
-	bug fix
	-	not working glyph substitution

[UTF-8 decoder capability and stress test]: http://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-test.txt

## 2012-12-19
-	version 0.2.3
-	bug fix
	-	wrong behavior of bce (ED, EL, DL, IL, ICH, DCH, ECH) (reported by IWAMOTO Kouichi ([@ttdoda]))

[@ttdoda]: http://doda.teraterm.org/whoami.xhtm

## 2014-03-07
-	version 0.2.4
-	add some parameters in conf.h
	- ROTATE: NORMAL or CLOCKWISE or COUNTER_CLOCKWISE or UPSIDE_DOWN
	- LAZY_DRAW (reduce drawing)
	- BACKGROUND_DRAW (after vt switch, continue drawing)

## 2014-03-24
-	version 0.2.5
-	remove ROTATE, LAZY_DRAW, BACKGROUND_DRAW options
-	add some environment options
		-	rotate: clockwise (cw), counter_clockwise (ccw), upside_down (ud)
		-	background drawing: wallpaper (wall)
		ex) $ YAFT="wall cw ccw ud" yaft
		-	framebuffer device:
		ex) $ FRAMEBUFFER="/dev/fb1" yaft
-	add OSC4/OSC104/OSC8900 (glyph width report)
		(ref: http://uobikiemukot.github.io/yaft/glyph_width_report.html)

## 2014-05-16
-	version 0.2.6
-	remove rotate option
-	support glyph width report rev.3
	(ref: http://uobikiemukot.github.io/yaft/glyph_width_report.html)
-	support DRCS/DRCSMMv1
-	bug fix
	-	never refresh cmap at vt switch (8bpp mode)

## 2014-05-18
-	version 0.2.7
-	support xim (yaftx)

## 2014-08-27
-	version 0.2.8
-	support sixel
-	support copy/paste (yaftx)
-	support NetBSD/OpenBSD wscons
-	refer SHELL environment variable
-	remove WALLPAPER option (use YAFT="wall" environment variable)
-	add man page

## 2015-07-26
-	version 0.2.8
-	rewrite framebuffer code
	-	support direct color
-	add some new options (conf.h)
	VT_CONTROL     : don't handle vt-switching (vt-switching only works in linux)
	FORCE_TEXT_MODE: need this option for switching from X to yaft
-	various bug fixes
	-	mkfont_bdf was broken (not affect SUBSTITUTE_WIDE/REPLACEMENT_CHAR)


x/x.h
/* See LICENSE for licence details. */
#include <X11/Xatom.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/keysym.h>

#define XK_NO_MOD UINT_MAX

enum { /* default terminal size 80x24 (8x16 dot font) */
	TERM_WIDTH  = 640,
	TERM_HEIGHT = 384,
};

struct keymap_t {
	KeySym keysym;
	unsigned int mask;
	char str[BUFSIZE];
};

const struct keymap_t keymap[] = {
	{XK_BackSpace, XK_NO_MOD, "\177"      },
	{XK_Up,        Mod1Mask,  "\033\033[A"},
	{XK_Up,        XK_NO_MOD, "\033[A"    },
	{XK_Down,      Mod1Mask,  "\033\033[B"},
	{XK_Down,      XK_NO_MOD, "\033[B"    },
	{XK_Right,     Mod1Mask,  "\033\033[C"},
	{XK_Right,     XK_NO_MOD, "\033[C"    },
	{XK_Left,      Mod1Mask,  "\033\033[D"},
	{XK_Left,      XK_NO_MOD, "\033[D"    },
	{XK_Begin,     XK_NO_MOD, "\033[G"    },
	{XK_Home,      XK_NO_MOD, "\033[1~"   },
	{XK_Insert,    XK_NO_MOD, "\033[2~"   },
	{XK_Delete,    XK_NO_MOD, "\033[3~"   },
	{XK_End,       XK_NO_MOD, "\033[4~"   },
	{XK_Prior,     XK_NO_MOD, "\033[5~"   },
	{XK_Next,      XK_NO_MOD, "\033[6~"   },
	{XK_F1,        XK_NO_MOD, "\033[[A"   },
	{XK_F2,        XK_NO_MOD, "\033[[B"   },
	{XK_F3,        XK_NO_MOD, "\033[[C"   },
	{XK_F4,        XK_NO_MOD, "\033[[D"   },
	{XK_F5,        XK_NO_MOD, "\033[[E"   },
	{XK_F6,        XK_NO_MOD, "\033[17~"  },
	{XK_F7,        XK_NO_MOD, "\033[18~"  },
	{XK_F8,        XK_NO_MOD, "\033[19~"  },
	{XK_F9,        XK_NO_MOD, "\033[20~"  },
	{XK_F10,       XK_NO_MOD, "\033[21~"  },
	{XK_F11,       XK_NO_MOD, "\033[23~"  },
	{XK_F12,       XK_NO_MOD, "\033[24~"  },
	{XK_F13,       XK_NO_MOD, "\033[25~"  },
	{XK_F14,       XK_NO_MOD, "\033[26~"  },
	{XK_F15,       XK_NO_MOD, "\033[28~"  },
	{XK_F16,       XK_NO_MOD, "\033[29~"  },
	{XK_F17,       XK_NO_MOD, "\033[31~"  },
	{XK_F18,       XK_NO_MOD, "\033[32~"  },
	{XK_F19,       XK_NO_MOD, "\033[33~"  },
	{XK_F20,       XK_NO_MOD, "\033[34~"  },
};

/* not assigned:
    kcbt=\E[Z,kmous=\E[M,kspd=^Z,
*/

struct xwindow_t {
	Display *display;
	Window window;
	Pixmap pixbuf;
	Colormap cmap;
	GC gc;
	int width, height;
	int screen;
	unsigned long color_palette[COLORS];
};

unsigned long color2pixel(struct xwindow_t *xw, uint32_t color)
{
	XColor xc;

	xc.red   = ((color >> 16) & bit_mask[8]) << 8;
	xc.green = ((color >>  8) & bit_mask[8]) << 8;
	xc.blue  = ((color >>  0) & bit_mask[8]) << 8;

	XAllocColor(xw->display, xw->cmap, &xc);
	return xc.pixel;
}

bool xw_init(struct xwindow_t *xw)
{
	XTextProperty xtext = {.value = (unsigned char *) "yaftx",
		.encoding = XA_STRING, .format = 8, .nitems = 5};

	if ((xw->display = XOpenDisplay(NULL)) == NULL) {
		logging(ERROR, "XOpenDisplay() failed\n");
		return false;
	}

	if (!XSupportsLocale())
		logging(WARN, "X does not support locale\n");

	if (XSetLocaleModifiers("") == NULL)
		logging(WARN, "cannot set locale modifiers\n");

	xw->cmap = DefaultColormap(xw->display, xw->screen);
	for (int i = 0; i < COLORS; i++)
		xw->color_palette[i] = color2pixel(xw, color_list[i]);

	xw->screen = DefaultScreen(xw->display);
	xw->window = XCreateSimpleWindow(xw->display, DefaultRootWindow(xw->display),
		0, 0, TERM_WIDTH, TERM_HEIGHT, 0, xw->color_palette[DEFAULT_FG], xw->color_palette[DEFAULT_BG]);
	XSetWMProperties(xw->display, xw->window, &xtext, NULL, NULL, 0, NULL, NULL, NULL); /* return void */

	xw->gc = XCreateGC(xw->display, xw->window, 0, NULL);
	XSetGraphicsExposures(xw->display, xw->gc, False);

	xw->width  = DisplayWidth(xw->display, xw->screen);
	xw->height = DisplayHeight(xw->display, xw->screen);
	xw->pixbuf = XCreatePixmap(xw->display, xw->window,
		xw->width, xw->height, XDefaultDepth(xw->display, xw->screen));

	XSelectInput(xw->display, xw->window,
		ExposureMask | KeyPressMask | StructureNotifyMask);

	XMapWindow(xw->display, xw->window);

	return true;
}

void xw_die(struct xwindow_t *xw)
{
	XFreeGC(xw->display, xw->gc);
	XFreePixmap(xw->display, xw->pixbuf);
	XDestroyWindow(xw->display, xw->window);
	XCloseDisplay(xw->display);
}

static inline void draw_sixel(struct xwindow_t *xw, int line, int col, struct cell_t *cellp)
{
	int w, h;
	uint32_t color = 0;

	for (h = 0; h < CELL_HEIGHT; h++) {
		for (w = 0; w < CELL_WIDTH; w++) {
			memcpy(&color, cellp->pixmap + BYTES_PER_PIXEL * (h * CELL_WIDTH + w), BYTES_PER_PIXEL);

			if (color_list[DEFAULT_BG] != color) {
				XSetForeground(xw->display, xw->gc, color2pixel(xw, color));
				XDrawPoint(xw->display, xw->pixbuf, xw->gc,
					col * CELL_WIDTH + w, line * CELL_HEIGHT + h);
			}
		}
	}
}

static inline void draw_line(struct xwindow_t *xw, struct terminal_t *term, int line)
{
	int bdf_padding, glyph_width, margin_right;
	int col, w, h;
	struct color_pair_t color_pair;
	struct cell_t *cellp;
	const struct glyph_t *glyphp;

	/* at first, fill all pixels of line in backgournd color */
	XSetForeground(xw->display, xw->gc, xw->color_palette[DEFAULT_BG]);
	XFillRectangle(xw->display, xw->pixbuf, xw->gc, 0, line * CELL_HEIGHT, term->width, CELL_HEIGHT);

	for (col = term->cols - 1; col >= 0; col--) {
		margin_right = (term->cols - 1 - col) * CELL_WIDTH;

		/* draw sixel pixmap */
		cellp = &term->cells[line * term->cols + col];
		if (cellp->has_pixmap) {
			draw_sixel(xw, line, col, cellp);
			continue;
		}

		/* get color and glyph */
		color_pair = cellp->color_pair;
		glyphp     = cellp->glyphp;

		/* check wide character or not */
		glyph_width = (cellp->width == HALF) ? CELL_WIDTH: CELL_WIDTH * 2;
		bdf_padding = my_ceil(glyph_width, BITS_PER_BYTE) * BITS_PER_BYTE - glyph_width;
		if (cellp->width == WIDE)
			bdf_padding += CELL_WIDTH;

		/* check cursor positon */
		if ((term->mode & MODE_CURSOR && line == term->cursor.y)
			&& (col == term->cursor.x
			|| (cellp->width == WIDE && (col + 1) == term->cursor.x)
			|| (cellp->width == NEXT_TO_WIDE && (col - 1) == term->cursor.x))) {
			color_pair.fg = DEFAULT_BG;
			color_pair.bg = (!vt_active && BACKGROUND_DRAW) ? PASSIVE_CURSOR_COLOR: ACTIVE_CURSOR_COLOR;
		}

		for (h = 0; h < CELL_HEIGHT; h++) {
			/* if UNDERLINE attribute on, swap bg/fg */
			if ((h == (CELL_HEIGHT - 1)) && (cellp->attribute & attr_mask[ATTR_UNDERLINE]))
				color_pair.bg = color_pair.fg;

			for (w = 0; w < CELL_WIDTH; w++) {
				/* set color palette */
				if (glyphp->bitmap[h] & (0x01 << (bdf_padding + w)))
					XSetForeground(xw->display, xw->gc, xw->color_palette[color_pair.fg]);
				else if (color_pair.bg != DEFAULT_BG)
					XSetForeground(xw->display, xw->gc, xw->color_palette[color_pair.bg]);
				else /* already draw */
					continue;

				/* update copy buffer */
				XDrawPoint(xw->display, xw->pixbuf, xw->gc,
					term->width - 1 - margin_right - w, line * CELL_HEIGHT + h);
			}
		}
	}

	term->line_dirty[line] = ((term->mode & MODE_CURSOR) && term->cursor.y == line) ? true: false;
}

void refresh(struct xwindow_t *xw, struct terminal_t *term)
{
	int line, update_from, update_to;

	if (term->mode & MODE_CURSOR)
		term->line_dirty[term->cursor.y] = true;

	update_from = update_to = -1;
	for (line = 0; line < term->lines; line++) {
		if (term->line_dirty[line]) {
			draw_line(xw, term, line);

			if (update_from == -1)
				update_from = update_to = line;
			else
				update_to = line;
		}
	}

	/* actual display update: vertical synchronizing */
	if (update_from != -1)
		XCopyArea(xw->display, xw->pixbuf, xw->window, xw->gc, 0, update_from * CELL_HEIGHT,
			term->width, (update_to - update_from + 1) * CELL_HEIGHT, 0, update_from * CELL_HEIGHT);
}
